src/full/Agda/Utils/Suffix.hs

```diff
@@ -45,37 +37,35 @@ fromSubscriptDigit d
 -- | Classification of identifier variants.
 
 data Suffix
-  = NoSuffix
-  | Prime     Int  -- ^ Identifier ends in @Int@ many primes.
+  = Prime     Int  -- ^ Identifier ends in @Int@ many primes.
   | Index     Int  -- ^ Identifier ends in number @Int@ (ordinary digits).
   | Subscript Int  -- ^ Identifier ends in number @Int@ (subscript digits).
 
 -- | Increase the suffix by one.  If no suffix yet, put a subscript @1@
 --   unless users do not want us to use any unicode.
 
 nextSuffix :: Suffix -> Suffix
-nextSuffix NoSuffix      = case subscriptAllowed of
-  UnicodeOk -> Subscript 1
-  AsciiOnly -> Index 1
 nextSuffix (Prime i)     = Prime $ i + 1
 nextSuffix (Index i)     = Index $ i + 1
 nextSuffix (Subscript i) = Subscript $ i + 1
 
 -- | Parse suffix.
 
-suffixView :: String -> (String, Suffix)
+suffixView :: String -> (String, Maybe Suffix)
 suffixView s
-    | (ps@(_:_), s') <- span (=='\'') rs         = (reverse s', Prime $ length ps)
-    | (ns@(_:_), s') <- span isDigit rs          = (reverse s', Index $ read $ reverse ns)
-    | (ns@(_:_), s') <- span isSubscriptDigit rs = (reverse s', Subscript $ read $
+    | (ps@(_:_), s') <- span (=='\'') rs         = (reverse s', Just $ Prime $ length ps)
+    | (ns@(_:_), s') <- span isDigit rs          = (reverse s', Just $ Index $ read $ reverse ns)
+    | (ns@(_:_), s') <- span isSubscriptDigit rs = (reverse s', Just $ Subscript $ read $
                                                       map fromSubscriptDigit $ reverse ns)
-    | otherwise                                  = (s, NoSuffix)
+    | otherwise                                  = (s, Nothing)
```

Comment:
By the way, this currently happens every time a new subscripted name is created; I'm not sure how often that occurs, but seems to be called from a few places:
  * `Agda.Syntax.Scope.Monde (freshConcreteName)`
  * `Agda.Syntax.Translation.AbstractToConcrete (chooseName, addSuffixConcrete)`
  * `Agda.Syntax.Translation.ReflectedToAbstract (withName)`

Recalling some ancient discussion in #3741, I wonder if the identifiers' string reversals might be causing some more GC churn than anticipated?

Reply:
Which reversals are you referring to?

Reply:
@nad Clearer with the full context: https://github.com/rwe/agda/blob/5937bce8d7fd952ef428c4fbd19db2ba8a124e4c/src/full/Agda/Utils/Suffix.hs#L54-L61

The way fresh names are generated is by appending or incrementing a suffix to the concrete name.

The way that is done is by first attempting to extract out any existing suffix of the concrete name in question, each time. That extraction done by reversing the string, finding "leading" suffix characters; and then reversing the resulting base name + suffix characters, then incrementing the suffix and checking for collision. It does that until no collision occurs.

Again, I'm not sure how often this happens, so it might not be a real concern. But it just my eye because I previously believed that after parsing the names were treated fairly opaquely, and that doesn't seem to be the case here.

Reply:
(This PR doesn't change that algorithm, by the way).

Reply:
Correction: `Agda.Syntax.Scope.Monad (freshConcreteName)` behaves differently than the others, in that it just appends a counted integer until no collision occurs, rather than parsing it out. The others in `ReflectedToAbstract` and `AbstractToConcrete` act as described, though, via `Agda.Syntax.Concrete.Name (nextName, firstNonTakenName)`.

Reply:
Ugh. Correction-to-correction: `Agda.Syntax.Scope.Monad` might do both. It appends an integer in `freshConcreteName`, but might do suffix parsing/appending in `tryResolveName`.